contracts/vaults/balancer/internal/pool/BalancerComposablePoolUtils.sol

```diff
@@ -0,0 +1,451 @@
+    /// @return primaryAmount primary token balance
+    function _getTimeWeightedPrimaryBalance(
+        BalancerComposablePoolContext memory poolContext,
+        ComposableOracleContext memory oracleContext,
+        StrategyContext memory strategyContext,
+        uint256 bptAmount,
+        bool validateOnly
+    ) internal view returns (uint256 primaryAmount) {
+        uint256 numTokens = poolContext.basePool.tokens.length;
+
+        for (uint256 i; i < numTokens; i++) {
+            if (i == poolContext.bptIndex) continue;
+            uint256 j = i + 1;
+            if (j == poolContext.bptIndex) j++;
+            if (j < numTokens) {
+                _checkPriceLimit(poolContext, oracleContext, strategyContext, i, j);
+            }
+
+            if (!validateOnly) {
+                primaryAmount += _convertToPrimary({
+                    poolContext: poolContext, 
+                    strategyContext: strategyContext,
+                    oracleContext: oracleContext,
+                    poolClaim: bptAmount,
+                    index: i
+                });
+            }
+        }
+    }
+
+    function _approveBalancerTokens(ComposablePoolContext memory poolContext, address bptSpender) internal {
+        for (uint256 i; i < poolContext.tokens.length; i++) {
+            IERC20(poolContext.tokens[i]).checkApprove(address(Deployments.BALANCER_VAULT), type(uint256).max);
+        }
+
+        // Allow BPT spender to pull BALANCER_POOL_TOKEN
+        poolContext.poolToken.checkApprove(bptSpender, type(uint256).max);
+    }
+
+    function _deposit(
+        BalancerComposablePoolContext memory poolContext,
+        ComposableOracleContext memory oracleContext,
+        StrategyContext memory strategyContext,
+        AuraStakingContext memory stakingContext,
+        uint256 deposit,
+        ComposableDepositParams memory params
+    ) internal returns (uint256 strategyTokensMinted) {
+
+        uint256[] memory amounts = _handleDepositTrades(poolContext, strategyContext, deposit, params);
+
+        uint256 bptMinted = _joinPoolAndStake({
+            poolContext: poolContext,
+            oracleContext: oracleContext,
+            strategyContext: strategyContext,
+            stakingContext: stakingContext,
+            amounts: amounts,
+            minBPT: params.minPoolClaim
+        });
+
+        strategyTokensMinted = strategyContext._mintStrategyTokens(bptMinted);
+    }
+
+    function _handleDepositTrades(
+        BalancerComposablePoolContext memory poolContext,
+        StrategyContext memory strategyContext,
+        uint256 deposit,
+        ComposableDepositParams memory params
+    ) private returns (uint256[] memory amounts) {
+        uint256 numTokens = poolContext.basePool.tokens.length;
+        amounts = new uint256[](poolContext.basePool.tokens.length);
+
+        if (params.depositTrades.length > 0) {
+            uint256 tradeIndex;
+            for (uint256 i; i < numTokens; i++) {
```

Comment:
This is a nit, but this structure seems a little wonky - you're iterating through both currencies in the pool and trades in the array. What if the trades don't line up with the currency ids by mistake? To me, it seems like it would be cleaner to just put the buyToken on the trade object and then just iterate through the trades.